test_app/app/controllers/application_controller.rb

```diff
@@ -0,0 +1,2 @@
+class ApplicationController < ActionController::Base
```

Comment:
What is this file about?

Reply:
# Application Controller

# What is Application Controller ( Summary )

The application controller is part of the model-view-controller and is also **the logical center of our application, it handles most of the business logic**. ( Sometimes we write some logic into model, services or helpers to prevent the controller from having too many codes and being hard to read..but this is another topic...). It coordinates the interaction between the user, the views, and the model. It is responsible for routing external requests to internal actions.

# What is a Controller?

Action Controller is the C in the MVC and MVC stands for Model-View-Controller, which is a popular web development design paradigm. 

- **Model layer**
    - The ***Model layer*** represents the domain model (such as Account, Product, Person, Post, etc.) and encapsulates the business logic specific to your application. In Rails, database-backed model classes are derived from `ActiveRecord::Base`. [Active Record](https://github.com/rails/rails/blob/main/activerecord/README.rdoc) allows you to present the data from database rows as objects and embellish these data objects with business logic methods. Although most Rails models are backed by a database, models can also be ordinary Ruby classes, or Ruby classes that implement a set of interfaces as provided by the [Active Model](https://github.com/rails/rails/blob/main/activemodel/README.rdoc) module.
- **View Layer**
    - The ***View layer*** is composed of "templates" that are responsible for providing appropriate representations of your application's resources. Templates can come in a variety of formats, but most view templates are HTML with embedded Ruby code (ERB files). Views are typically rendered to generate a controller response or to generate the body of an email. In Rails, View generation is handled by [Action View](https://github.com/rails/rails/blob/main/actionview/README.rdoc).
- **Controller Layer**
    - The ***Controller layer*** is responsible for handling incoming HTTP requests and providing a suitable response. Usually, this means returning HTML, but Rails controllers can also generate XML, JSON, PDFs, mobile-specific views, and more. Controllers load and manipulate models, and render view templates in order to generate the appropriate HTTP response. In Rails, incoming requests are routed by Action Dispatch to an appropriate controller, and controller classes are derived from `ActionController::Base`. Action Dispatch and Action Controller are bundled together in [Action Pack](https://github.com/rails/rails/blob/main/actionpack/README.rdoc).

Here we are focusing on the controller layer, a controller may be thought of as a link between models and views. It makes model data available to the view so that it can display it to the user, as well as save or update user data in the model. After the router has determined which controller to use for a request, the controller is responsible for making sense of the request and producing the appropriate output. Luckily, Action Controller does most of the groundwork for you and uses smart conventions to make this as straightforward as possible.

## Action Pack - The magic behind the controller

### Introduction of Action Pack

Action Pack is a framework for handling and responding to web requests. It provides mechanisms for **routing** (mapping request URLs to actions), defining **controllers** that implement actions, and generating responses. In short, Action Pack provides the controller layer in the MVC paradigm.

It consists of several modules:

- Action Dispatch, which parses information about the web request, handles routing as defined by the user, and does advanced processing related to HTTP such as MIME-type negotiation ( Examples of MIME types are: **text/html for normal web pages**. text/plain for plain text. application/octet-stream meaning “download this file” ), decoding parameters in POST, PATCH, or PUT bodies, handling HTTP caching logic, cookies and sessions.
- Action Controller, which provides a base controller class that can be subclassed to implement filters and actions to handle requests. The result of an action is typically content generated from views.

With the Ruby on Rails framework, users only directly interface with the Action Controller module. Necessary Action Dispatch functionality is activated by default and Action View rendering is implicitly triggered by Action Controller. However, these modules are designed to function on their own and can be used outside of Rails.

### **Dispatching Requests to Controllers**

At its simplest, a web application accepts an incoming request from a browser, processes it, and sends a response.

Rails provide two ways to define how to route a request:

- Define a direct mapping of URLs to actions based on pattern matching, requirements, and conditions(the comprehensive way).
- Define routes based on resources, such as the models that you define(the convenient way).
- we can freely mix and match the two approaches, because the convenient way is built on the comprehensive way.

**In below I will be trying to depict the process of dispatching requests to controllers.**

We use routes.rb to set up the reference to our controller. However, the `RouteSet` is the actual class that acts as the entry point for route configuration in a Rails application. It’s most famous for the `#draw` method, which we’ve just used in `routes.rb` ( The Rails.application.routes is an instance of Routes set and the draw function block that wraps your route definitions is required to establish the scope for the router DSL and must not be deleted )

```ruby
# Example config/routes.rb file provided in https://guides.rubyonrails.org/routing.html
Rails.application.routes.draw do
  resources :brands, only: [:index, :show] do
    resources :products, only: [:index, :show]
  end

  resource :basket, only: [:show, :update, :destroy]

  resolve("Basket") { route_for(:basket) }
end

# The draw function is defined in rails/actionpack/lib/action_dispatch/routing/route_set.rb
def draw(&block)
  clear! unless @disable_clear_and_finalize
  eval_block(block)
  finalize! unless @disable_clear_and_finalize
  nil
end
```

At runtime, `RouteSet` is responsible for coordinating the entire operation that we’re about to dive into: it receives the incoming web request, and collaborates with the objects below to determine how the request makes it into our application code.

When Rails is booting, [a new `RouteSet` gets instantiated](https://github.com/rails/rails/blob/v6.0.0.beta2/railties/lib/rails/engine.rb#L533-L537). It evaluates the contents of the routes file and builds up a `RouteSet`.

Because `RouteSet` is the source of truth for all available endpoints in our application, it’s also first in line to receive a request from the outside world, after passing through Rack and various middleware. That’s right, this humble class buried in `ActionPack` is the Walmart greeter of our application, ready with a smile and a wave as soon as a request comes through the door.

In order for `RouteSet` to accept the request after it’s travelled through Rack and any middleware, it needs to implement Rack’s interface, which is as simple as implementing `call`:

### **REST(Representational State Transfer)**

REST, or REpresentational State Transfer, is an architectural style for providing standards between computer systems on the web, making it easier for systems to communicate with each other. REST-compliant systems, often called RESTful systems, are characterized by how they are stateless ( the request are independent, the request should not involve any information from the previous request, there is no order ) and separate the concerns of client and server (the code on the client-side can be changed at any time without affecting the operation of the server, and the code on the server-side can be changed without affecting the operation of the client.). 

In REST, we use a simple set of verbs to operate on a rich set of nouns. If we’re using HTTP, the verbs correspond to HTTP methods (GET, PUT, PATCH, POST, and DELETE, typically). The nouns are the resources in our application. We name those resources using URLs. We can use *rails routes* to get the list of defined routes, the Verb column shows the request method, URI shows the URL to call the controller, and the Controller#Action column shows which method to use when the controller instance has been created.

```ruby
                                  Prefix Verb                        URI Pattern                                                                                    Controller#Action
           rails_postmark_inbound_emails POST   /rails/action_mailbox/postmark/inbound_emails(.:format)                                           action_mailbox/ingresses/postmark/inbound_emails#create
              rails_relay_inbound_emails POST   /rails/action_mailbox/relay/inbound_emails(.:format)                                              action_mailbox/ingresses/relay/inbound_emails#create
           rails_sendgrid_inbound_emails POST   /rails/action_mailbox/sendgrid/inbound_emails(.:format)                                           action_mailbox/ingresses/sendgrid/inbound_emails#create
     rails_mandrill_inbound_health_check GET    /rails/action_mailbox/mandrill/inbound_emails(.:format)                                           action_mailbox/ingresses/mandrill/inbound_emails#health_check
           rails_mandrill_inbound_emails POST   /rails/action_mailbox/mandrill/inbound_emails(.:format)                                           action_mailbox/ingresses/mandrill/inbound_emails#create
            rails_mailgun_inbound_emails POST   /rails/action_mailbox/mailgun/inbound_emails/mime(.:format)                                       action_mailbox/ingresses/mailgun/inbound_emails#create
          rails_conductor_inbound_emails GET    /rails/conductor/action_mailbox/inbound_emails(.:format)                                          rails/conductor/action_mailbox/inbound_emails#index
                                         POST   /rails/conductor/action_mailbox/inbound_emails(.:format)                                          rails/conductor/action_mailbox/inbound_emails#create
       new_rails_conductor_inbound_email GET    /rails/conductor/action_mailbox/inbound_emails/new(.:format)                                      rails/conductor/action_mailbox/inbound_emails#new
      edit_rails_conductor_inbound_email GET    /rails/conductor/action_mailbox/inbound_emails/:id/edit(.:format)                                 rails/conductor/action_mailbox/inbound_emails#edit
           rails_conductor_inbound_email GET    /rails/conductor/action_mailbox/inbound_emails/:id(.:format)                                      rails/conductor/action_mailbox/inbound_emails#show
                                         PATCH  /rails/conductor/action_mailbox/inbound_emails/:id(.:format)                                      rails/conductor/action_mailbox/inbound_emails#update
                                         PUT    /rails/conductor/action_mailbox/inbound_emails/:id(.:format)                                      rails/conductor/action_mailbox/inbound_emails#update
                                         DELETE /rails/conductor/action_mailbox/inbound_emails/:id(.:format)                                      rails/conductor/action_mailbox/inbound_emails#destroy
new_rails_conductor_inbound_email_source GET    /rails/conductor/action_mailbox/inbound_emails/sources/new(.:format)                              rails/conductor/action_mailbox/inbound_emails/sources#new
   rails_conductor_inbound_email_sources POST   /rails/conductor/action_mailbox/inbound_emails/sources(.:format)                                  rails/conductor/action_mailbox/inbound_emails/sources#create
   rails_conductor_inbound_email_reroute POST   /rails/conductor/action_mailbox/:inbound_email_id/reroute(.:format)                               rails/conductor/action_mailbox/reroutes#create
                      rails_service_blob GET    /rails/active_storage/blobs/redirect/:signed_id/*filename(.:format)                               active_storage/blobs/redirect#show
                rails_service_blob_proxy GET    /rails/active_storage/blobs/proxy/:signed_id/*filename(.:format)                                  active_storage/blobs/proxy#show
                                         GET    /rails/active_storage/blobs/:signed_id/*filename(.:format)                                        active_storage/blobs/redirect#show
               rails_blob_representation GET    /rails/active_storage/representations/redirect/:signed_blob_id/:variation_key/*filename(.:format) active_storage/representations/redirect#show
         rails_blob_representation_proxy GET    /rails/active_storage/representations/proxy/:signed_blob_id/:variation_key/*filename(.:format)    active_storage/representations/proxy#show
                                         GET    /rails/active_storage/representations/:signed_blob_id/:variation_key/*filename(.:format)          active_storage/representations/redirect#show
                      rails_disk_service GET    /rails/active_storage/disk/:encoded_key/*filename(.:format)                                       active_storage/disk#show
               update_rails_disk_service PUT    /rails/active_storage/disk/:encoded_token(.:format)                                               active_storage/disk#update
                    rails_direct_uploads POST   /rails/active_storage/direct_uploads(.:format)                                                    active_storage/direct_uploads#create
```

### **Processing of Requests**

When a controller object processes a request, it looks for a public instance method with the same name as the incoming action. If it finds one, that method is invoked. If it doesn’t find one and the controller implements `method_missing()`, that method is called, passing in the action name as the first parameter and an empty argument list as the second. If no method can be called, the controller looks for a template named after the current controller and action. If found, this template is rendered directly. If none of these things happens, an `AbstractController::ActionNotFound` error is generated.

### **Controller Environment**

The controller sets up the environment for actions (and, by extension, for the views that they invoke). Many of these methods provide direct access to information contained in the URL or request.

- action_name: the name of the action currently being processed.
- cookies: the cookies associated with the request, and setting values into this object stores cookies on the browser when the response is sent.
- headers: a hash of HTTP headers that will be used in the response.
- params: a hash-like object containing request parameters (along with pseudopa- rameters generated during routing).
- request: the incoming request object.
- response: the response object, filled in during the handling of the request, normally managed for you by Rails.
- session: a hash-like object representing the current session data.
- logger

Some of **the attributes that a request object has:**

- request_method: returns request method ( :delete, :get, :head, :post, or :put ). In the case where the method was overridden by a middleware (for instance, if a HEAD request was converted to a `[GET](https://api.rubyonrails.org/classes/ActionDispatch/Request.html#method-i-GET)`, or if a *method parameter was used to determine the method the application should use), this method returns the overridden value, not the original. 
The following code will show how the request_method works, we first convert and assign our methods into variable according to the protocols and store inside a object with key-value pair. Afterwards we pass our request to the request_method, it will look up the method inside the object, if the method with same name was found then it will assign the method name to the @request_method instance variable. If not found then raise a ActionController::UnknownHttpMethod error.*

```ruby
#rails/actionpack/lib/action_dispatch/http/request.rb

# List of HTTP request methods from the following RFCs:
# Hypertext Transfer Protocol -- HTTP/1.1 (https://www.ietf.org/rfc/rfc2616.txt)
# HTTP Extensions for Distributed Authoring -- WEBDAV (https://www.ietf.org/rfc/rfc2518.txt)
# Versioning Extensions to WebDAV (https://www.ietf.org/rfc/rfc3253.txt)
# Ordered Collections Protocol (WebDAV) (https://www.ietf.org/rfc/rfc3648.txt)
# Web Distributed Authoring and Versioning (WebDAV) Access Control Protocol (https://www.ietf.org/rfc/rfc3744.txt)
# Web Distributed Authoring and Versioning (WebDAV) SEARCH (https://www.ietf.org/rfc/rfc5323.txt)
# Calendar Extensions to WebDAV (https://www.ietf.org/rfc/rfc4791.txt)
# PATCH Method for HTTP (https://www.ietf.org/rfc/rfc5789.txt)

RFC2616 = %w(OPTIONS GET HEAD POST PUT DELETE TRACE CONNECT)
RFC2518 = %w(PROPFIND PROPPATCH MKCOL COPY MOVE LOCK UNLOCK)
RFC3253 = %w(VERSION-CONTROL REPORT CHECKOUT CHECKIN UNCHECKOUT MKWORKSPACE UPDATE LABEL MERGE BASELINE-CONTROL MKACTIVITY)
RFC3648 = %w(ORDERPATCH)
RFC3744 = %w(ACL)
RFC5323 = %w(SEARCH)
RFC4791 = %w(MKCALENDAR)
RFC5789 = %w(PATCH)

HTTP_METHODS = RFC2616 + RFC2518 + RFC3253 + RFC3648 + RFC3744 + RFC5323 + RFC4791 + RFC5789

HTTP_METHOD_LOOKUP = {}

# Populate the HTTP method lookup cache.
HTTP_METHODS.each { |method|
  HTTP_METHOD_LOOKUP[method] = method.underscore.to_sym
}

def request_method
  @request_method ||= check_method(super)
end

private
  def check_method(name)
    HTTP_METHOD_LOOKUP[name] || raise(ActionController::UnknownHttpMethod, "#{name}, accepted HTTP methods are #{HTTP_METHODS[0...-1].join(', ')}, and #{HTTP_METHODS[-1]}")
    name
  end
end
```

- method: Returns the original value of the environment's REQUEST_METHOD, even if it was overridden by middleware. See `[request_method](https://api.rubyonrails.org/classes/ActionDispatch/Request.html#method-i-request_method)`
 for more information. Usually, it returns the same value as request_method except for :head, which it returns as :get instead because those two method are functionally equivalent from an application point of view. The following code shows the way to do it.

```ruby
def method
  @method ||= check_method(get_header("rack.methodoverride.original_method") || get_header("REQUEST_METHOD"))
end
```

- delete?, get?, head?, post? and put? return true or false based on the request method.
- xml_http_request? and xhr? return true if the “X-Requested-With” header contains “XMLHttpRequest” (case-insensitive), which may need to be manually added depending on the choice of JavaScript libraries and frameworks.  [SEE HERE TO KNOW WHAT IS XMLHttpRequest](https://www.notion.so/XMLHttpRequest-e47f8e0c2b774c27ba8beec6cd418d02)
- headers(), which is the complete set of HTTP headers.

```ruby
# rails/actionpack/lib/action_dispatch/http/request.rb
# Provides access to the request's HTTP headers, for example:
# request.headers["Content-Type"] # => "text/plain"
def headers
  @headers ||= Http::Headers.new(self)
end
```

- body(), which is the request body as an I/O stream.
- content_length(), which is the number of bytes purported to be in the body.

Ruby leverages a gem named Rack to provide much of this functionality.  Rack::Request provides a convenient interface to a Rack environment. It is stateless, the environment `env` passed to the constructor will be directly modified.

```ruby
req =Rack::Request.new(env)
req.post?
req.params["data"]
```

### **Responding to the User**

Part of the controller’s job is to respond to the user. There are basically four ways of doing this:

- render a template as the view taking information provided by the controller and using it to generate a response to the browser. A template is a file that defines the content of a response for our application. Rails supports three template formats out of the box:
    - erb, which is embedded Ruby code (typically with HTML).support
    - builder, a more programmatic way of constructing XML content.
    - RJS, which generates JavaScript.
- return a string directly to the browser without invoking a view, this is fairly rare but can be used to send error notifications.
- return nothing to the browser, this is sometimes used when responding to an Ajax request.
- send other data to the client (something other than HTML). This is typically a download of some kind (perhaps a PDF docu- ment or a file’s contents).

A controller always responds to the user exactly one time per request. This means that only one call to a `render()`, `redirect_to()`, or `[head](https://edgeapi.rubyonrails.org/classes/ActionController/Head.html#method-i-head)`method in the processing of any request. (A `DoubleRenderError` exception will be thrown on the second render if we have one.). From the controller's point of view, there are three ways to create an HTTP response:

- `render()`: the heart of all rendering in Rails. It takes a hash of options that tell it what to render and how to render it.

```ruby
# Normalizes arguments, options and then delegates render_to_body and
# sticks the result in <tt>self.response_body</tt>.
def render(*args, &block)
  options = _normalize_render(*args, &block)
  rendered_body = render_to_body(options)
  if options[:html]
    _set_html_content_type
  else
    _set_rendered_content_type rendered_format
  end
  _set_vary_header
  self.response_body = rendered_body
end
```

- `[head](https://edgeapi.rubyonrails.org/classes/ActionController/Head.html#method-i-head)` : ****head(status, options = {})**** Returns a response that has no content (merely headers). The options argument is interpreted to be a hash of header names and values. This allows you to easily return a response that consists only of significant headers:

```ruby
def head(status, options = {})
  if status.is_a?(Hash)
    raise ArgumentError, "#{status.inspect} is not a valid value for `status`."
  end

  status ||= :ok

  location = options.delete(:location)
  content_type = options.delete(:content_type)

  options.each do |key, value|
    headers[key.to_s.split(/[-_]/).each { |v| v[0] = v[0].upcase }.join("-")] = value.to_s
  end

  self.status = status
  self.location = url_for(location) if location

  if include_content?(response_code)
    unless self.media_type
      self.content_type = content_type || (Mime[formats.first] if formats) || Mime[:html]
    end

    response.charset = false
  end

  self.response_body = ""

  true
end
```

- `redirect_to()`: redirect to an action in a given controller, to a URL, or to the previous page.

```ruby
def redirect_to(options = {}, response_options = {})
  raise ActionControllerError.new("Cannot redirect to nil!") unless options
  raise AbstractController::DoubleRenderError if response_body

  allow_other_host = response_options.delete(:allow_other_host) { _allow_other_host }

  self.status        = _extract_redirect_to_status(options, response_options)
  self.location      = _enforce_open_redirect_protection(_compute_redirect_to_location(request, options), allow_other_host: allow_other_host)
  self.response_body = ""
end
```

# Controller Naming Convention

The naming convention of controllers in Rails favors pluralization of the last word in the controller's name, although it is not strictly required (e.g. `ApplicationController`). For example, `ClientsController` is preferable to `ClientController`, `SiteAdminsController` is preferable to `SiteAdminController` or `SitesAdminsController`, and so on.

Following this convention will allow you to use the default route generators (e.g. `resources`, etc) without needing to qualify each `:path` or `:controller`, and will keep named route helpers' usage consistent throughout your application. See [Layouts and Rendering Guide](https://guides.rubyonrails.org/layouts_and_rendering.html) for more details.

Application Controller is practically the class that every other controller in your application is going to inherit from (although this is not mandatory by any means).

# Methods and Actions

A controller is a Ruby class. We can have many controllers for different request and response handling purpose and has methods on their own. Those controllers are usually inherited from a base controller ( the default is ApplicationController ). This gives you one class to configure things such as request forgery protection and filtering of sensitive request parameters.

When our application receives a request, the routing will decide which controller and action ( The action here is referred to any public methods defined inside the controller ) to call, Rails will create an instance of that controller and run the method with the same name as the action.

```ruby
class **ClientsController** **<** ApplicationController
  def **new**
  end
end
```

As an example, if a user goes to `/clients/new` in your application to add a new client, Rails will create an instance of `ClientsController` and call its `new` method. Note that the empty method from the example above would work just fine because Rails will by default render the `new.html.erb` view unless the action says otherwise. By creating a new `Client`, the `new` method can make a `@client` instance variable accessible in the view:

```ruby
def **new**
	@client **=** Client.**new**
end
```

The actual rendering is done by nested classes of the module `[ActionView::Template::Handlers](https://api.rubyonrails.org/v7.0.2.3/classes/ActionView/Template/Handlers.html)`. This guide does not dig into that process, but it's important to know register an object that knows how to handle template files with the given [extensions](https://apidock.com/rails/ActionView/Template/Handlers/extensions/class). This can be used to implement new template types. The handler must respond to :call, which will be passed the template and should return the rendered template as a [String](https://apidock.com/rails/String).

# Best Practices

Only public methods are callable as actions. It is a best practice to lower the visibility of methods (with `private` or `protected`) which are not intended to be actions, like auxiliary methods or filters. By default, only the ApplicationController in a Rails application inherits from `ActionController::Base`. All other controllers inherit from ApplicationController. This gives you one class to configure things such as request forgery protection and filtering of sensitive request parameters.

# Sample Controller

A sample controller could look like this:

```ruby
class PostsController < ApplicationController
def index
	@posts = Post.all
end

def create 
  @post = Post.create params[:post]
    redirect_to posts_path
  end
end
```

Actions, by default, render a template in the `app/views` directory corresponding to the name of the controller and action after executing code in the action. For example, the `index` action of the PostsController would render the template `app/views/posts/index.html.erb` by default after populating the `@posts` instance variable. Unlike index, the create action will not render a template. After performing its main purpose (creating a new post), it initiates a redirect instead. This redirect works by returning an external `302 Moved` HTTP response that takes the user to the index action.

These two methods represent the two basic action archetypes used in Action Controllers: Get-and-show and do-and-redirect. Most actions are variations on these themes.